### PR TITLE
Introduce Targetables and expand target coverage

### PR DESCRIPTION
**Pull request by zetlen**
### Features
- Add [Targetables](https://github.com/magento/pwa-studio/blob/zetlen/targetables/pwa-devdocs/_drafts/targetables-explainer.md), a new helper library for powerful Targets
- Reimplement core Targets using Targetables
- Add an automatic API generator to Peregrine that exposes all hooks and talons in Targets

Excerpt from the explainer at https://github.com/magento/pwa-studio/blob/zetlen/targetables/pwa-devdocs/_drafts/targetables-explainer.md follows:

> # Targetables: Fluent tools for making your code extensible with Targets
> 
> Targetables are handy objects which represent source files in your project or library. When paired with Buildpack Targets, Targetables make it simple to customize PWA modules on the fly.
> 
> ## Introduction
> 
> You need a lot of Targets to make a PWA Studio project powerfully extensible. Enhancing the storefront through Targets means changing source code in highly specific ways during the build process. A Targetable for a source file provides methods for making pinpoint changes to the code in that file.
> 
> ```js
> targetables.esModule('@my/library/Button.js', {
>   publish() {
>     const focusClasses = this
>       .addImport('focusClasses from "src/a11y/focus-classes.css"');
> 
>     this
>       .setJSXProps('<button>', { tabIndex: '0' })
>       .addJSXClassName('<button>', `${focusClasses}.focusRing`)
>   }
> })
> ```
> 
> The above modifies the source of `Button.js` right before Webpack pulls it into the build. It's always on the fly, never saving those changes to `Button.js` on disk.

### Other changes

- Moved `ReactRefreshWebpackPlugin` to PWADevServer, so it only runs in watch mode. A plain build in dev mode caused an error.
- Switched from using `babel-loader` to a simple custom Babel loader. That loader adds the ability for Babel plugins to report errors and warnings that Webpack will log. Otherwise it works exactly the same.
- Fixed a bad idea in BuildBusPlugin, which was to queue log messages. Now it just logs them immediately.
- Moved the default routes in `venia-ui` out of the intercept itself and into a `defaultRoutes.json` file.
- Other very small bugfixes I discovered along the way.

## Related Issue

Closes PWA-875.

### Verification Stakeholders
The whole @magento/pwa-studio-team 
Community leaders @larsroettig and @Jordaneisenburger 
Community participants
### Verification Steps

1. Go to the [Demo section of the explainer](https://github.com/magento/pwa-studio/blob/zetlen/targetables/pwa-devdocs/_drafts/targetables-explainer.md#demo). 
1. Copy and paste the code there into `venia-concept/local-intercept.js` as instructed.
1. Run `yarn watch:venia`
1. Load the Venia site in your browser.
1. Observe the changes made by Targetables.
1. Repeat for production mode.

## Screenshots / Screen Captures (if appropriate)
image

Look, I was playing around. It works, okay?

## Checklist

- [x] I have added tests to cover my changes, if necessary.
* I have updated the documentation accordingly, if necessary.

## Notes

- The HookInterceptorSet reads the filesystem to build Peregrine APIs. This means that if someone adds, removes, or moves a while, those snapshots will break. This is on purpose; it tells the developer that changes have occurred in the public API! It may be necessary to modify or subclass HookInterceptorSet for compatibility in the future.

- @jcalcaben will need to decide how to document the talon and hook targets that are autogenerated.

- TargetableSet will probably need more and more utility methods. Arguably they don't go there, but it works today.

- I tried making the JSX methods in TargetableReactComponent more general-purpose and powerful by using [Babel templates](https://babeljs.io/docs/en/next/babel-template.html). Two main things went wrong:

  - Their API requires a lot of introspection, and doesn't tolerate any unknown properties.
  - Syntactic placeholders aren't compatible with being in JSXText position, so they broke in nonintuitive ways.

  This is worth more investigation by someone. Babel templates, or even callbacks, would make it much easier to create new operations. As it is, the `BabelModifyJSXPlugin` is pretty large, and each common method has to be made separately.

- I recommend that someone follow this up with a TargetableCSSModule that lets you do PostCSS transforms, to demonstrate that Targetables aren't just for JS.

### CODE REVIEW

File: pwa-devdocs/_drafts/targetables-explainer.md [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+If you're working on a PWA extension, Buildpack restricts your intercepts so they can only modify files in your own extension package. But Targetables are still really useful in your extension! You can add Targets to permit third-party intercepts of your own extension, and implement them concisely with Targetables.
+
+```js
+targetables.esModule('@my/library/Button.js', {
+  async publish(myTargets) {
+    const classnames = await myTargets.buttonClassnames.promise([]);
+    classnames.forEach(name => this.addJSXClassName('<button>', name));
+  }
+});
+```
```
Comment (jcalcaben) [inline pwa-devdocs/_drafts/targetables-explainer.md:35]: An explanation for this codeblock would be useful

File: packages/pwa-buildpack/lib/WebpackTools/targetables/TargetableModule.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+     * @returns TransformRequest[]
+     */
+    flush() {
+        return this._queuedTransforms.splice(0, this._queuedTransforms.length);
+    }
+    /**
+     * Insert text into the module contents, immediately following the location
+     * of the search string if it is found.
+     *
+     * @param {string} search - Text string in the module code to place the new content after.
```
Comment (jcalcaben) [inline packages/pwa-buildpack/lib/WebpackTools/targetables/TargetableModule.js:51]: I don't see this in the param list?

File: packages/pwa-buildpack/lib/WebpackTools/targetables/TargetableModule.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            after,
+            insert,
+            ...options
+        });
+    }
+    /**
+     * Insert text into the module contents, immediately before the location
+     * of the search string if it is found.
+     *
+     * @param {string} search - Text string in the module code to place the new content before.
```
Comment (jcalcaben) [inline packages/pwa-buildpack/lib/WebpackTools/targetables/TargetableModule.js:68]:
```suggestion
     * @param {string} before - Text string in the module code to place the new content before.
```

File: pwa-devdocs/_drafts/targetables-explainer.md [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  })
+}
+```
+
+Now, Freddy's module can add nav items using a higher-level concept:
+
+**@freddy/pwa-chat/intercept.js**
+```js
+module.exports = targets => {
+  targets.own.navBar.tap(bar => {
```
Comment (jimbo) [inline pwa-devdocs/_drafts/targetables-explainer.md:111]:
```suggestion
  targets.of('@example/pwa-nav').navBar.tap(bar => {
```

File: packages/peregrine/lib/targets/HookInterceptorSet.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
+const path = require('path');
+const packageRoot = path.resolve(__dirname, '../..');
+const glob = require('fast-glob');
+const Trackable = require('@magento/pwa-buildpack/lib/BuildBus/Trackable');
```
[the remark this answers is not in the document]
Comment (zetlen) [reply packages/peregrine/lib/targets/HookInterceptorSet.js:4]: Yes, Trackable is a utility class for hierarchical logging. It works--you can set `DEBUG=BuildBus` to see a whole heap of output--but it's not really finished. In normal circumstances all of its methods are noops and it does nothing.

File: packages/peregrine/lib/targets/peregrine-declare.js [hunk 6]
```diff
@@ -18,33 +47,35 @@ module.exports = targets => {
...
          * @example <caption>Wrap the `useApp()` hook  with a logging extension</caption>
          *
-         * talonsTarget.tap(talonWrapperConfig => {
-         *   talonWrapperConfig.App.useApp.wrapWith('@my-extensions/log-wrapper');
+         * talonsTarget.tap(talons => {
+         *   talons.App.useApp.wrapWith('@my-extensions/log-wrapper');
          * })
          */
-        talons: new targets.types.Sync(['talons'])
+        talons: new targets.types.AsyncSeries(['talons'])
```
Comment (zetlen) [inline packages/peregrine/lib/targets/peregrine-declare.js:66]: Note that changing a target from Sync to Async is a backwards-compatible change for interceptors. They can still use `tap()` with a synchronous interceptor.

**Comment by larsroettig** [comment]
@zetlen `DEBUG_PROJECT_CREATION=1 pwa-studio/packages/create-pwa/bin/create-pwa` is broken with *yarn*
```
$ webpack-dev-server --progress --color --env.mode development
SyntaxError: /home/roettigl/PhpstormProjects/slack/target/node_modules/@magento/venia-ui/lib/defaultRoutes.json: Unexpected token < in JSON at position 8
    at JSON.parse (<anonymous>)
    at Object.Module._extensions..json (internal/modules/cjs/loader.js:797:27)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at makeRoutesTarget (/home/roettigl/PhpstormProjects/slack/target/node_modules/@magento/venia-ui/lib/targets/makeRoutesTarget.js:8:26)
    at Object.module.exports.veniaTargets [as intercept] (/home/roettigl/PhpstormProjects/slack/target/node_modules/@magento/venia-ui/lib/targets/venia-ui-intercept.js:20:5)
    at pertaining.forEach.dep (/home/roettigl/PhpstormProjects/slack/target/node_modules/@magento/pwa-buildpack/lib/BuildBus/BuildBus.js:198:23)
    at Array.forEach (<anonymous>)
    at BuildBus.runPhase (/home/roettigl/PhpstormProjects/slack/target/node_modules/@magento/pwa-buildpack/lib/BuildBus/BuildBus.js:186:20)
    at BuildBus.init (/home/roettigl/PhpstormProjects/slack/target/node_modules/@magento/pwa-buildpack/lib/BuildBus/BuildBus.js:167:14)
    at configureWebpack (/home/roettigl/PhpstormProjects/slack/target/node_modules/@magento/pwa-buildpack/lib/WebpackTools/configureWebpack/configureWebpack.js:130:9)
Done in 6.23s.
```

File: packages/pwa-buildpack/lib/WebpackTools/targetables/__tests__/TargetableSet.spec.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        it('caches and polymorphically reuses only one module of one type per file', () => {
+            const file = '@magento/fakedep/esModule1';
+            const esModule1 = targetable.esModule(file);
+            expect(targetable.esModule(file)).toBe(esModule1);
+            expect(targetable.module(file)).toBe(esModule1);
+            expect(() =>
+                targetable.reactComponent(file)
+            ).toThrowErrorMatchingSnapshot();
+        });
+        describe('automatically intercepts transformModules and send its requests, running a "publish" method first if it is supplied', () => {
```
Comment (jimbo) [inline packages/pwa-buildpack/lib/WebpackTools/targetables/__tests__/TargetableSet.spec.js:63]:
```suggestion
        describe('automatically intercepts transformModules and sends its requests, running a "publish" method first if it is supplied', () => {
```

**Review by jimbo — COMMENTED** [review]
Okay, code looks good. I'll commit a couple of fixups in a bit here.

The approach makes sense to me so far. Next steps are to check the branch out and try the new APIs out.

File: packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/JSXSnippetParser.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
+const parseMethodOpts = {
+    plugins: ['syntax-jsx']
+};
+const elementPatterns = [
+    /^<>.+<\/>$/, // React fragments: <>{...}</>
+    /<.+\/\s*>$/, // Self-closing tags: <foo{...}/>
+    /^<([^\s>]+).+<\/\1>$/ // Tags that open and close: <foo{...}>{...}</foo>
+];
+const openingElementPattern = /^<.+>$/;
```
[the remark this answers is not in the document]
Comment (zetlen) [reply packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/JSXSnippetParser.js:9]: This is to enable the "element" matcher, the first argument to all the TargetableReactComponent JSX operations. It allows, e.g.:

```js
component.appendJSX(`<BrowserRouter>`, `<Route {...foo}>`);
```

Or 

```js
component.appendJSX(`BrowserRouter`, `<Route {...foo}>`);
```

Without this, you'd always have to write the first argument as a self-closing element `<BrowserRouter />` ...even if it isn't self-closing in the actual code you're querying.

Babel won't parse just any excerpt of JS in isolation; without lexer state, it doesn't know what kind of token is legal. So it won't parse just a JSX opening element. Therefore we do a little bit of regex first, to add angle brackets if they're missing and then to add a self-closing slash.

**Review by jimbo — DISMISSED** [review]

**Comment by dpatil-magento** [comment]
@jimbo Please check failed tests and also while trying to run verification steps I see error on `yarn build` or `yarn watch:venia` 
```

yarn run v1.22.4
$ yarn venia run build
$ yarn workspace @magento/venia-concept run build
$ yarn run build:prod
$ yarn run clean && webpack --no-progress --env.mode production
$ rimraf dist
TypeError: dep[phase] is not a function
    at /pwa-studio/packages/pwa-buildpack/lib/BuildBus/BuildBus.js:198:23
    at Array.forEach (<anonymous>)
    at BuildBus.runPhase (/pwa-studio/packages/pwa-buildpack/lib/BuildBus/BuildBus.js:186:20)
    at BuildBus.init (/pwa-studio/packages/pwa-buildpack/lib/BuildBus/BuildBus.js:167:14)
    at configureWebpack (/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/configureWebpack/configureWebpack.js:130:9)
    at async module.exports (/pwa-studio/packages/venia-concept/webpack.config.js:24:20)
error Command failed with exit code 1
```

**Comment by larsroettig** [comment]
@dpatil-magento tested this branch on the weekend don't see this issue for watch and build also i tested scaffolfing

**Comment by dpatil-magento** [comment]
Thanks @larsroettig, its just the [Demo](https://github.com/magento/pwa-studio/blob/zetlen/targetables/pwa-devdocs/_drafts/targetables-explainer.md#demo) code is missing `module.exports = localIntercept;`. After adding this, it worked for me.

**Review by larsroettig — APPROVED** [review]

**Review by dpatil-magento — APPROVED** [review]
Approved latest commits

**Review by dpatil-magento — APPROVED** [review]

**Pull request merged**